### PR TITLE
Proxy range requests

### PR DESCRIPTION
**Pull request by golenkovm**
Issue #335 and #294 
* Implement get_filesize_by_contenthash() file system method.
* Introduce unit test case for manager class.
* Tweak get_header() method and add unit tests.
* Allow associative arrays to be passed to get_header() method.
* Add a patch for Moodle 3.3/Totara 12.
* Introduce Proxy range request feature.
* Add test range request on setting form.
* Add unit tests covering Proxy range request methods.

### CODE REVIEW

File: classes/local/store/object_file_system.php [hunk 1]
```diff
@@ -898,4 +913,68 @@ public function get_filesize_by_contenthash($contenthash) {
...
+     */
+    public function serve_range_request_via_presigned_url($contenthash, $filesize, $ranges) {
+        $url = $this->externalclient->generate_presigned_url($contenthash, headers_list());
+        $headers = array(
+            'HTTP/1.1 206 Partial Content',
+            'Content-Length: '.$ranges->length,
+            'Range: bytes=' . $ranges->rangefrom . '-' . $ranges->rangeto,
+        );
+
+        $curl = curl_init();
```
Comment (brendanheywood) [inline classes/local/store/object_file_system.php:959]: please avoid using the raw php curl libs, use the moodle lib:

$curl = new curl();

This is generally preferred, but if you are already using guzzle for some reasons then its ok as long as the moodle proxy options are honored

File: classes/local/store/object_file_system.php [hunk 2]
```diff
@@ -898,4 +913,68 @@ public function get_filesize_by_contenthash($contenthash) {
...
+        curl_setopt($curl, CURLOPT_CONNECTTIMEOUT, 15);
+        curl_setopt($curl, CURLOPT_TIMEOUT, 15);
+        $content = curl_exec($curl);
+        curl_close($curl);
+
+        if ($content === false) {
+            return false;
+        } else {
+            header('HTTP/1.1 206 Partial Content');
+            header('Content-Range: bytes ' . $ranges->rangefrom . '-' . $ranges->rangeto . '/' . $filesize);
```
Comment (brendanheywood) [inline classes/local/store/object_file_system.php:973]: Technically it is possible for the range that was requested to not be the range that was returned. Can you please grab these headers from the curl response and pass them through as-is

File: classes/local/store/object_file_system.php [hunk 3]
```diff
@@ -421,6 +421,13 @@ public function xsendfile_file(stored_file $file): bool {
 
             return $this->redirect_to_presigned_url($contenthash, headers_list());
         }
+
+        if ($this->externalclient->support_presigned_urls()) {
+            $ranges = $this->get_valid_http_ranges($file->get_filesize());
+            if ($file->get_filesize() > 2097152 && $ranges) {
```
Comment (brendanheywood) [inline classes/local/store/object_file_system.php:427]: can you make a const for this magic number and add a comment explaining it

Comment (brendanheywood) [reply classes/local/store/object_file_system.php:427]: probably point to the php max temp / aws sdk comments

File: classes/local/store/object_file_system.php [hunk 4]
```diff
@@ -443,6 +450,14 @@ public function xsendfile($contenthash) {
 
             return $this->redirect_to_presigned_url($contenthash, $headers);
         }
+
+        if ($this->externalclient->support_presigned_urls()) {
+            $filesize = $this->get_filesize_by_contenthash($contenthash);
+            $ranges = $this->get_valid_http_ranges($filesize);
+            if ($filesize > 2097152 && $ranges) {
+                $this->serve_range_request_via_presigned_url($contenthash, $file->get_filesize(), $ranges);
```
Comment (brendanheywood) [inline classes/local/store/object_file_system.php:458]: is this a bug?

                $this->serve_range_request_via_presigned_url($contenthash, $file->get_filesize(), $ranges);

vs

                $this->serve_range_request_via_presigned_url($contenthash, $filesize, $ranges);

File: classes/local/store/object_file_system.php (hunk 1 above)
Comment (golenkovm) [reply classes/local/store/object_file_system.php]: `new curl();` has handy features indeed. Thanks.
Done.

File: classes/local/store/object_file_system.php (hunk 2 above)
Comment (golenkovm) [reply classes/local/store/object_file_system.php]: Done

File: classes/local/store/object_file_system.php (hunk 3 above)
Comment (golenkovm) [reply classes/local/store/object_file_system.php]: Made a constant and added a comment with a link.
Didn't find how to get this php max temp on the fly. 
Looks like it's 2MB by default, but can be overriden via code. And there is no way to get current `maxmemory`.

File: classes/local/store/object_file_system.php (hunk 4 above)
Comment (golenkovm) [reply classes/local/store/object_file_system.php]: Yep. Thanks. Fixed

File: classes/local/store/object_file_system.php [hunk 5]
```diff
@@ -891,4 +901,105 @@ private function exec_command($command) {
...
+     * @param  stored_file $file    The file to send
+     * @param  object      $ranges  Object with rangefrom, rangeto and length properties.
+     * @return false if couldn't get data.
+     * @throws \coding_exception
+     */
+    public function serve_range_request(stored_file $file, $ranges) {
+        $response = $this->curl_range_request_to_presigned_url($file->get_contenthash(), $ranges, headers_list());
+        $httpcode = manager::get_header($response['responseheaders'], 'HTTP/1.1');
+        if ($response['content'] == '' || $httpcode != '206 Partial Content') {
+            debugging('Range request failed with HTTP code: ' . $httpcode . '. Details: ' . $response['content']);
```
Comment (brendanheywood) [inline classes/local/store/object_file_system.php:955]: can you log the failed curl url here please

File: classes/local/store/object_file_system.php [hunk 6]
```diff
@@ -421,6 +429,13 @@ public function xsendfile_file(stored_file $file): bool {
 
             return $this->redirect_to_presigned_url($contenthash, headers_list());
         }
+
+        if ($this->externalclient->support_presigned_urls()) {
```
Comment (brendanheywood) [inline classes/local/store/object_file_system.php:433]: as discussed move this logic into serve_range_request() which will be in each store

Comment (golenkovm) [reply classes/local/store/object_file_system.php]: Done

File: classes/local/store/object_file_system.php (hunk 5 above)
Comment (golenkovm) [reply classes/local/store/object_file_system.php]: Done

File: classes/local/store/s3/client.php [hunk 7]
```diff
@@ -37,6 +37,14 @@
 define('AWS_CAN_DELETE_OBJECT', 2);
 
 class client extends object_client_base {
+
+    /**
+     * @var int A predefined limit of data stored.
+     * When hit, php://temp will use a temporary file.
+     * Reference: https://www.php.net/manual/en/wrappers.php.php
```
Comment (brendanheywood) [inline classes/local/store/s3/client.php:44]: Can you please link to this:

https://github.com/catalyst/moodle-local_aws/blob/master/sdk/Aws/S3/StreamWrapper.php#L19-L25

File: classes/local/store/s3/client.php [hunk 8]
```diff
@@ -635,4 +643,71 @@ public function define_client_check_sdk($config) {
...
+        }
+
+        $response = $this->curl_range_request_to_presigned_url($file->get_contenthash(), $ranges, headers_list());
+        $httpcode = manager::get_header($response['responseheaders'], 'HTTP/1.1');
+
+        if ($response['content'] != '' && $httpcode == '206 Partial Content') {
+            header('HTTP/1.1 206 Partial Content');
+            header('Accept-Ranges: bytes');
+            $contentrange = manager::get_header($response['responseheaders'], 'Content-Range');
+            if ($contentrange !== '') {
```
Comment (brendanheywood) [inline classes/local/store/s3/client.php:669]: I think these 2 if statements could be simpler, it should always trust the headers from the s3 response. If those headers don't exist then don't fall back to $contentrange, just treat that the same as the other types of failures if the code is not 206. If they don't have these headers they are a totally invalid response

https://tools.ietf.org/html/rfc7233#section-4.1

> If a single part is being transferred, the server generating the 206 response MUST generate a Content-Range header field

File: classes/local/store/object_file_system.php [hunk 9]
```diff
@@ -773,7 +773,7 @@ public function presigned_url_should_redirect_file($file) {
      * @throws \dml_exception
      */
     public function presigned_url_should_redirect($contenthash, $headers = array()) {
-        global $DB, $_SERVER;
+        global $_SERVER;
```
Comment (brendanheywood) [inline classes/local/store/object_file_system.php:784]: not needed?

Comment (golenkovm) [reply classes/local/store/object_file_system.php:784]: Yeah

File: classes/local/store/s3/client.php [hunk 10]
```diff
@@ -635,4 +643,63 @@ public function define_client_check_sdk($config) {
...
+     * Proxy range request.
+     *
+     * @param  \stored_file $file    The file to send
+     * @param  object       $ranges  Object with rangefrom, rangeto and length properties.
+     * @return false If couldn't get data.
+     * @throws \coding_exception
+     */
+    public function proxy_range_request(\stored_file $file, $ranges) {
+        // Do not serve files with size less than 2MB or if the feature is disabled.
+        if ($file->get_filesize() < self::MAX_TEMP_LIMIT || empty($this->config->proxyrangerequests)) {
```
Comment (brendanheywood) [inline classes/local/store/s3/client.php:658]: can you swap the order of this so if the config is off it exist earlier

get_filesize can do some db / api calls in same cases

**Pull request merged**